### ISSUE TITLE
Anomaly Score is calculated wrongly 

### ISSUE THREAD

**Issue by Harrymsys** [opendistro-for-elasticsearch/anomaly-detection#223]
**Describe the bug**
Anomaly Score is calculated wrongly for the following scenario,

**To Reproduce**

-> I am running this project in local and the two plugins are deployed in the elastic search 
-> I created the same anomaly detector as described here for testing and understanding: https://opendistro.github.io/for-elasticsearch-docs/docs/ad/api/#create-anomaly-detector 

```
"feature_attributes": [
    {
      "feature_name": "total_order",
      "feature_enabled": true,
      "aggregation_query": {
        "total_order": {
          "sum": {
            "field": "value"
          }
        }
      }
    }
  ]
```
Also, my detector interval is 1 minute as same as given in the documentation

-> I start the detector and it is running
-> Parallelly using my own test code, I am populating the "order" index in the elastic search which populates for every 3 seconds as a 'value' field between 1 to 10 randomly and when it reaches to 100 divisible, I am updating the huge value in the 'value' column in 'order' index to check the anomaly. 

```
                Map<String, Object> dataMap = new HashMap<String, Object>();
		dataMap.put("timestamp", DATE_FORMAT.format(new Date()));
		dataMap.put("value", random.nextInt(10));
		if(atomicValue.get() % 100 == 0) {
			dataMap.put("value", atomicValue.get());
		}
```

which means the value will be populated like 

Example : 

![image](https://user-images.githubusercontent.com/37929362/92515083-a6dbdc00-f230-11ea-9654-706710e42bd5.png)

But anomaly score is not increasing when the value is increased, Please find the below output 

**Expected behavior**
When the sum of the 'value' in order index is modified. expecting the change in the anomaly score but the anomaly score is not calculated as expected, Please find the screenshot

**Screenshots**

1) Creating the detector 

![image](https://user-images.githubusercontent.com/37929362/92515742-b60f5980-f231-11ea-89c7-fb52f2789dbc.png)

2) Starting the detector 

![image](https://user-images.githubusercontent.com/37929362/92515808-d50deb80-f231-11ea-90e6-416e8c3f2fd5.png)

3) Parallelly Order index is populated using my own test code 

4) Model is created 

![image](https://user-images.githubusercontent.com/37929362/92515510-5e70ee00-f231-11ea-9cb5-b094e936b291.png)

5) PF the anomaly detector results

```
"hits": [
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "2REQb3QBhcHTa30MZQ4i",
                "_version": 1,
                "_seq_no": 9,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 5.088753089094179,
                    "execution_start_time": 1599591178599,
                    "data_end_time": 1599591118599,
                    "confidence": 0.8238839646482641,
                    "data_start_time": 1599591058599,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 401
                        }
                    ],
                    "execution_end_time": 1599591179553,
                    "anomaly_grade": 0.01234567901233773
                },
                "sort": [
                    1599591179553
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "wxEPb3QBhcHTa30MeQ5-",
                "_version": 1,
                "_seq_no": 8,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 4.534258135489182,
                    "execution_start_time": 1599591118600,
                    "data_end_time": 1599591058600,
                    "confidence": 0.8237597742294861,
                    "data_start_time": 1599590998600,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 94
                        }
                    ],
                    "execution_end_time": 1599591119230,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599591119230
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "phEOb3QBhcHTa30MjQ7J",
                "_version": 1,
                "_seq_no": 7,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 4.636067976770329,
                    "execution_start_time": 1599591058600,
                    "data_end_time": 1599590998600,
                    "confidence": 0.823635410425475,
                    "data_start_time": 1599590938600,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 75
                        }
                    ],
                    "execution_end_time": 1599591058889,
                    "anomaly_grade": 0.007444168734482831
                },
                "sort": [
                    1599591058889
                ]
            },
```

We can see the total_order is calculated for every one minute,

Could someone please explain the anomaly score compare to the sum of the field 'value' [total_order]? 
Because 

when the total_order = 75, Anomaly Score = 4.636067976770329
when the total_order = 94, Anomaly Score = 4.534258135489182
when the total_order = 401, Anomaly Score = 5.088753089094179

**Because as per my knowledge the anomaly score is not calculated as expected. when the sum of the field 'value' is increased drastically (94 to 401) I have expected a huge increase in the anomaly score but it is NOT.** 

Any help is appreciated.

Thanks,  
Harry

**Comment by wnbts** [comment]
Hello Harry, thanks for using the product and providing feedback!

1. Users are expected to use anomaly grades in the results. Anomaly scores are only for internal use and should not be used directly. In this example, when total order is 401, the anomaly grade is greater than 0, indicating a possible anomaly.

2. The anomaly score does not grow linearly with the input. It is an approximate and normalized. So it is not expected to increase as significantly as the input. Rather its rank relative to other anomaly scores matters. In this example, the score when total order is 401 is larger than the other scores, which is an expected behavior.

**Comment by Harrymsys** [comment]
> Hello Harry, thanks for using the product and providing feedback!
> 
> 1. Users are expected to use anomaly grades in the results. Anomaly scores are only for internal use and should not be used directly. In this example, when total order is 401, the anomaly grade is greater than 0, indicating a possible anomaly.
> 2. The anomaly score does not grow linearly with the input. It is an approximate and normalized. So it is not expected to increase as significantly as the input. Rather its rank relative to other anomaly scores matters. In this example, the score when total order is 401 is larger than the other scores, which is an expected behavior.

@wnbts , appreciate for your immediate reply. 

Even the anomaly grade is not reliable, Could you check the following recent result, I have a few values which has more than 1K but the anomaly grade is 0, Could you please explain me this. 

```
{
    "took": 7,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 81,
            "relation": "eq"
        },
        "max_score": null,
        "hits": [
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "EhFRb3QBhcHTa30MZhUI",
                "_version": 1,
                "_seq_no": 80,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 3.056780848862962,
                    "execution_start_time": 1599595438600,
                    "data_end_time": 1599595378600,
                    "confidence": 0.8322797650950773,
                    "data_start_time": 1599595318600,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 97
                        }
                    ],
                    "execution_end_time": 1599595439624,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595439624
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "-RFQb3QBhcHTa30MehQu",
                "_version": 1,
                "_seq_no": 79,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 3.04237589782296,
                    "execution_start_time": 1599595378598,
                    "data_end_time": 1599595318598,
                    "confidence": 0.8321670170333029,
                    "data_start_time": 1599595258598,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 1786
                        }
                    ],
                    "execution_end_time": 1599595379246,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595379246
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "4xFPb3QBhcHTa30MjhRG",
                "_version": 1,
                "_seq_no": 78,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 2.3098666824448326,
                    "execution_start_time": 1599595318600,
                    "data_end_time": 1599595258600,
                    "confidence": 0.832054118956252,
                    "data_start_time": 1599595198600,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 95
                        }
                    ],
                    "execution_end_time": 1599595318853,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595318853
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "zhFOb3QBhcHTa30MphSP",
                "_version": 1,
                "_seq_no": 77,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 2.2035178386286765,
                    "execution_start_time": 1599595258598,
                    "data_end_time": 1599595198598,
                    "confidence": 0.8319410705648899,
                    "data_start_time": 1599595138598,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 76
                        }
                    ],
                    "execution_end_time": 1599595259535,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595259535
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "txFNb3QBhcHTa30MuhTh",
                "_version": 1,
                "_seq_no": 76,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 2.8785583624048896,
                    "execution_start_time": 1599595198600,
                    "data_end_time": 1599595138600,
                    "confidence": 0.8318278715593869,
                    "data_start_time": 1599595078600,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 85
                        }
                    ],
                    "execution_end_time": 1599595199201,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595199201
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "oRFMb3QBhcHTa30MzhSh",
                "_version": 1,
                "_seq_no": 75,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 3.047753838327571,
                    "execution_start_time": 1599595138599,
                    "data_end_time": 1599595078599,
                    "confidence": 0.8317145216391151,
                    "data_start_time": 1599595018599,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 67
                        }
                    ],
                    "execution_end_time": 1599595138721,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595138721
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "iRFLb3QBhcHTa30M5hS7",
                "_version": 1,
                "_seq_no": 74,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 3.15850719139804,
                    "execution_start_time": 1599595078598,
                    "data_end_time": 1599595018598,
                    "confidence": 0.8316010205026458,
                    "data_start_time": 1599594958598,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 1698
                        }
                    ],
                    "execution_end_time": 1599595079355,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595079355
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "cxFKb3QBhcHTa30M-hTU",
                "_version": 1,
                "_seq_no": 73,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 2.335833405199727,
                    "execution_start_time": 1599595018599,
                    "data_end_time": 1599594958599,
                    "confidence": 0.8314873678477472,
                    "data_start_time": 1599594898599,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 117
                        }
                    ],
                    "execution_end_time": 1599595018964,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599595018964
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "XhFKb3QBhcHTa30MEhTY",
                "_version": 1,
                "_seq_no": 72,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 2.2274035527874654,
                    "execution_start_time": 1599594958600,
                    "data_end_time": 1599594898600,
                    "confidence": 0.8313735633713821,
                    "data_start_time": 1599594838600,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 95
                        }
                    ],
                    "execution_end_time": 1599594959576,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599594959576
                ]
            },
            {
                "_index": ".opendistro-anomaly-results-history-2020.09.08-1",
                "_type": "_doc",
                "_id": "RxFJb3QBhcHTa30MJhTy",
                "_version": 1,
                "_seq_no": 71,
                "_primary_term": 1,
                "_score": null,
                "_source": {
                    "detector_id": "MhEJb3QBhcHTa30M5g77",
                    "anomaly_score": 2.9980361762200953,
                    "execution_start_time": 1599594898600,
                    "data_end_time": 1599594838600,
                    "confidence": 0.8312596067697037,
                    "data_start_time": 1599594778600,
                    "feature_data": [
                        {
                            "feature_id": "LhEJb3QBhcHTa30M5g73",
                            "feature_name": "total_order",
                            "data": 88
                        }
                    ],
                    "execution_end_time": 1599594899186,
                    "anomaly_grade": 0
                },
                "sort": [
                    1599594899186
                ]
            }
        ]
    }
}
```

**Comment by wnbts** [comment]
Since anomalies should by definition be uncommon, to put the number of anomalies under control, the model will learn 99.5% data as normal and anomaly rate is 0.5%. So if you inject the anomalies too frequently, they are likely to be learned as normal data.

**Comment by Harrymsys** [comment]
@wnbts , Consider the following case

0 to 1 min : Total Value = 30 
1 to 2 min : Total Value = 20 
2 to 3 min : Total Value = 35 
3 to 4 min : Total Value = 29 
**4 to 5 min : Total Value = 100** 
5 to 6 min : Total Value = 50 
6 to 7 min : Total Value = 40 
7 to 8 min : Total Value = 65 
8 to 9 min : Total Value = 39 
**9 to 10 min : Total Value = 200** 

Similarly, for every 5th minute, I am increasing the total value as +100 
I was expecting the anomaly_grade to be increased from 0 to 1 for every 5th-minute entry but it is NOT 

It is fundamentally wrong, right? What is your thought on this?

**Comment by wnbts** [comment]
For the case you gave, here's my thoughts/opinions, please feel free to discuss about your needs.

1.
The most core message from is that anomaly detection is meant for raising rare data, not necessarily "wrong" data :)

Is values 100/200 "wrong"? It depends. Without clear rules, right/wrong is a subject matter. Maybe, but maybe the value 20 is wrong to some. After all, 20 orders is worse for business than 200.

Is value 100/200 anomalous? It depends. From our experience, it's an even more subject matter. If the new values are like 98, 102, 214, etc, or a new value is like 999,999, it would be harder to call 100/200 anomalous, wouldn't it?

Is top/bottom 20% rare? It depends. Imagine in real life you are watching the data and find a "rare anomaly" every five minute (20%), how to process them in time?

In addition, if you know what is wrong data vs right data, you may consider using thresholding or learning (classification) methods.

2.
Accuracy consists of precision and recall. Increasing anomaly rate is likely to increase recall but decrease precision. So increasing anomaly rate to 20% might produce false positives and hurt accuracy. Currently, this product uses a fixed anomaly rate 0.5%. Note with 1440 1-min samples in a day, 0.5% anomaly rate produces roughly 7 anomalies daily, which should be plenty for human to investigate. What anomaly rate is appropriate for your data and business needs and is it practical to consume an increased number of anomalies? (288 anomalies a day for 20% anomaly rate) If it sounds reasonable, we are open to making this available for customization.

**Comment by ylwu-amzn** [comment]
@Harrymsys  will close this issue as no updates for more than 2 months, feel free to reopen if you still have question.
